modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java

```diff
@@ -1785,31 +1818,48 @@ private void connectInCommonMode() throws SQLException {
             }
         }
 
-        handleConnectExceptions(exceptions);
+        handleConnectExceptions(exceptions, singleIo != null);
     }
 
     /**
      * Prepare and throw general {@code SQLException} with all specified exceptions as suppressed items.
      *
      * @param exceptions Exceptions list.
+     * @param disconnected Flag indicating that a disconnection from the server has occurred.
      * @throws SQLException Umbrella exception.
      */
-    private void handleConnectExceptions(List<Exception> exceptions) throws SQLException {
+    private void handleConnectExceptions(List<Exception> exceptions, boolean disconnected) throws SQLException {
         if (connCnt.get() == 0 && exceptions != null) {
-            close();
+            if (disconnected) {
+                closeStatementsOnDisconnect(null);
+            }
+
+            if (!disconnected || txIo != null) {
+                // We need to implicitly release all resources if the initial connection attempt fails or
+                // if the transaction context exists, because in this case we cannot reestablish the connection.
+                close();
+            }
 
             if (exceptions.size() == 1) {
                 Exception ex = exceptions.get(0);
 
-                if (ex instanceof SQLException)
+                if (ex instanceof SQLException) {
+                    if (disconnected) {
+                        throw new SQLException(
+                            ex.getMessage(), ((SQLException)ex).getSQLState(), new JdbcThinDisconnectException(ex));
+                    }
+
```

Comment:
Do we need all these exception wrapping? 

Maybe a "if (exceptions.size() == 1) " branch can be dropped? 
Or left it as is and add it something like that
```
if (disconnected) {
   e = new JdbcThinDisconnectException(null);
    for (Exception ex : exceptions)
                e.addSuppressed(ex);
    throw e;
}
```

Reply:
> Do we need all these exception wrapping?
:tired_face:  :thinking: 

The task description states that we need to somehow signal the client that a shutdown has occurred, but he can repeat the operation at his own risk.

But after some thinking I decided to drop `JdbcThinDisconnectException`, because it looks like the client can check the SQL error state to understand what happened, something like this.
```
   catch(SQLException e) {
     if (!conn.isClosed() && "08006".equals(ex.getSQLState()))
       // means that disconnection happened but connection can be re-used (operation can be retried
   }
```
states described here: https://www.gridgain.com/docs/latest/developers-guide/SQL/JDBC/error-codes
(`08006` mean `I/O error during communication`) :thinking: 

WDYT?